### PR TITLE
Backward pass ac

### PR DESCRIPTION
**Pull request by paryxyt**
This PR introduces changes to the portion of the partitioner responsible for activation checkpointing (AC). In past work, changes were introduced where by activations were determined for check-pointing based on a knapsack -optimization formulation, and these were solved by either ILP, or a greedy heuristic, or a dynamic programming approach.
We identified an opportunity to further improve upon these heuristics. The main insight has been that an AC policy that introduces "long" recomputation chains are likely to be inefficienct, and drive memory-peaks upward due to the costs involved with rematerialization. These can be avoided by "breaking" these chains via saving a small set of strategically chosen nodes. In this diff, we introduce an approach for doing so. The nodes to be saved can be determined by solving an Integer Linear Program (ILP).
image

 
In the above, x, a binary decision-variable corresponds to save vs recompute decisions. Z_i corresponds to the memory needed to compute node i. If any of node i's dependencies (say node j) is not saved, it will need to be rematerialized. Thus the memory cost of materializing node i is the sum of m_i (the memory cost of node i) and the rematerialization costs of all the dependencies. The decision variable y is the peak memory cost. One nuance worth noting is that the above formulation involves a product of decision variables, Z_jx_j, which is "non-linear". To address this, we have to use a "linearization" trick that is standard in the Integer Linear Programming literature. This involves introducing a new variable Qj = Z_jx_j, and some addition constraints on Qj.
The diff defines a new class called LongRecomputationChains. Within this class we formulate the ILP and the solve_ILP() method solves the same. We also have additional code that is needed to form the appropriate data objects that are needed. The output is used to modify the saved_values_set in order to reduce the memory usage during backward pass. Additionally, the config.py file has been updated to include a new parameter that controls the option to choose the backward-pass-aware option and also the aggressiveness of the backward pass-aware AC.
Test Plan
ran on PyTorch benchmark "BERT_pytorch"

### CODE REVIEW

**Comment by paryxyt** [comment]
@Chillee @jansel please review?

File: torch/_functorch/partitioners.py [hunk 1]
```diff
@@ -1494,21 +1495,50 @@ def dp_knapsack(
...
         return greedy_knapsack(memory, runtimes, max_memory)
     elif SOLVER == "ilp":
+        memory = list(memory.values())
         return ilp_knapsack(memory, runtimes, max_memory)
     elif SOLVER == "dp":
+        memory = list(memory.values())
         return dp_knapsack(memory, runtimes, max_memory)
+    elif SOLVER == "backward_pass_aware":
+        try:
+            print("----- solving ILP based approach ----")
```
Comment (jansel) [inline torch/_functorch/partitioners.py:1525]: Use logging not prints

File: torch/_functorch/partitioners.py [hunk 2]
```diff
@@ -1494,21 +1495,50 @@ def dp_knapsack(
...
+                max_memory,
+                node_info,
+                all_recomputable_banned_nodes,
+            )
+            return [0.0, saved_node_idx, recomp_node_idx]
+        except Exception as e:
+            print(f"ILP based approach failed with error: {e}")
+            print("----- solving DP based approach ----")
+            memory = list(memory.values())
+            return dp_knapsack(memory, runtimes, max_memory)
```
Comment (jansel) [inline torch/_functorch/partitioners.py:1538]: Why does this fail?  We should fix tese failures rather than silently ignoring them.

**Review by jansel — CHANGES_REQUESTED** [review]
Fix test failures and do a full run on the perf dashboard.

**Pull request closed without merging**